### PR TITLE
[Kernel] [Triton] [AMD] Add Triton implementation of awq_dequantize

### PR DESCRIPTION
**Pull request by rasmith**
This PR adds awq_dequantize_triton, an implementation of awq_dequantize, which is already implemented in CUDA, as a fallback for when the CUDA implementation cannot be used. PR created per request of Joe Shajwari based on his conversation with Woosuk.

A Triton based implementation of awq_gemm will be coming soon to make it possible to use AWQ in some basic level when CUDA implementation cannot be used. 
**BEFORE SUBMITTING, PLEASE READ THE CHECKLIST BELOW AND FILL IN THE DESCRIPTION ABOVE**

---

<details>

<summary><b> PR Checklist (Click to Expand) </b></summary>

<p>Thank you for your contribution to vLLM! Before submitting the pull request, please ensure the PR meets the following criteria. This helps vLLM maintain the code quality and improve the efficiency of the review process.</p>

<h3>PR Title and Classification</h3>
<p>Only specific types of PRs will be reviewed. The PR title is prefixed appropriately to indicate the type of change. Please use one of the following:</p>
<ul>
    <li><code>[Bugfix]</code> for bug fixes.</li>
    <li><code>[CI/Build]</code> for build or continuous integration improvements.</li>
    <li><code>[Doc]</code> for documentation fixes and improvements.</li>
    <li><code>[Model]</code> for adding a new model or improving an existing model. Model name should appear in the title.</li>
    <li><code>[Frontend]</code> For changes on the vLLM frontend (e.g., OpenAI API server, <code>LLM</code> class, etc.) </li>
    <li><code>[Kernel]</code> for changes affecting CUDA kernels or other compute kernels.</li>
    <li><code>[Core]</code> for changes in the core vLLM logic (e.g., <code>LLMEngine</code>, <code>AsyncLLMEngine</code>, <code>Scheduler</code>, etc.)</li>
    <li><code>[Hardware][Vendor]</code> for hardware-specific changes. Vendor name should appear in the prefix (e.g., <code>[Hardware][AMD]</code>).</li>
    <li><code>[Misc]</code> for PRs that do not fit the above categories. Please use this sparingly.</li>
</ul>
<p><strong>Note:</strong> If the PR spans more than one category, please include all relevant prefixes.</p>

<h3>Code Quality</h3>

<p>The PR need to meet the following code quality standards:</p>

<ul>
    <li>We adhere to <a href="https://google.github.io/styleguide/pyguide.html">Google Python style guide</a> and <a href="https://google.github.io/styleguide/cppguide.html">Google C++ style guide</a>.</li>
    <li>Pass all linter checks. Please use <a href="https://github.com/vllm-project/vllm/blob/main/format.sh"><code>format.sh</code></a> to format your code.</li>
    <li>The code need to be well-documented to ensure future contributors can easily understand the code.</li>
    <li>Include sufficient tests to ensure the project to stay correct and robust. This includes both unit tests and integration tests.</li>
    <li>Please add documentation to <code>docs/source/</code> if the PR modifies the user-facing behaviors of vLLM. It helps vLLM user understand and utilize the new features or changes.</li>
</ul>

<h3>Notes for Large Changes</h3>
<p>Please keep the changes as concise as possible. For major architectural changes (>500 LOC excluding kernel/data/config/test), we would expect a GitHub issue (RFC) discussing the technical design and justification. Otherwise, we will tag it with <code>rfc-required</code> and might not go through the PR.</p>

<h3>What to Expect for the Reviews</h3>

<p>The goal of the vLLM team is to be a <i>transparent reviewing machine</i>. We would like to make the review process transparent and efficient and make sure no contributor feel confused or frustrated. However, the vLLM team is small, so we need to prioritize some PRs over others. Here is what you can expect from the review process: </p>

<ul>
    <li> After the PR is submitted, the PR will be assigned to a reviewer. Every reviewer will pick up the PRs based on their expertise and availability.</li>
    <li> After the PR is assigned, the reviewer will provide status update every 2-3 days. If the PR is not reviewed within 7 days, please feel free to ping the reviewer or the vLLM team.</li>
    <li> After the review, the reviewer will put an <code> action-required</code> label on the PR if there are changes required. The contributor should address the comments and ping the reviewer to re-review the PR.</li>
    <li> Please respond to all comments within a reasonable time frame. If a comment isn't clear or you disagree with a suggestion, feel free to ask for clarification or discuss the suggestion.
 </li>
</ul>

<h3>Thank You</h3>

<p> Finally, thank you for taking the time to read these guidelines and for your interest in contributing to vLLM. Your contributions make vLLM a great tool for everyone! </p>

</details>

### CODE REVIEW

**Comment by casper-hansen** [comment]
@robertgshaw2-neuralmagic could you take a look? I think Triton implementations could be interesting to expand access to other hardware.

File: vllm/model_executor/layers/quantization/awq_triton.py [hunk 1]
```diff
@@ -0,0 +1,272 @@
...
+        print(f"Torch result:iweights_torch = {iweights_torch}")
+
+    if use_torch and use_triton:
+        diff = iweights_torch - iweights_triton
+        error = torch.sum(torch.sqrt(diff * diff))
+        print(f"error = {error}")
+
+
+if __name__ == '__main__':
+    main()
```
Comment (mgoin) [inline vllm/model_executor/layers/quantization/awq_triton.py:272]: Could you convert this to a separate test file?

File: vllm/model_executor/layers/quantization/awq_triton.py [hunk 2]
```diff
@@ -0,0 +1,272 @@
...
+    )
+    reverse_order_tensor = reverse_order_tensor.view(-1, 32 // bits)
+    reverse_order_tensor = reverse_order_tensor[:, AWQ_REVERSE_ORDER]
+    reverse_order_tensor = reverse_order_tensor.view(-1)
+
+    t = t[:, reverse_order_tensor] & 0xF
+    return t
+
+
+# qweightss [R     , C // 8], int32
```
Comment (mgoin) [inline vllm/model_executor/layers/quantization/awq_triton.py:174]:
```suggestion
# qweights  [R     , C // 8], int32
```

File: vllm/_custom_ops.py [hunk 3]
```diff
@@ -9,6 +9,11 @@
 
 logger = init_logger(__name__)
 
+# NOTE: 7/26/24: Use Triton implementation of AWQ functions. Currently, only
+#       awq_dequantize is implemented, but awq_gemm coming soon.
+
+use_awq_triton = False
```
Comment (mgoin) [inline vllm/_custom_ops.py:15]: How does this get set? Should this be an environment variable or based on hardware available?

File: vllm/model_executor/layers/quantization/awq_triton.py (hunk 1 above)
Comment (rasmith) [reply vllm/model_executor/layers/quantization/awq_triton.py:272]: Thank you for reviewing, but I've moved it over to this one https://github.com/vllm-project/vllm/pull/7386 , instead of two PRs, I'm going going to have one.

**Comment by rasmith** [comment]
@mgoin Please review this one if you could: https://github.com/vllm-project/vllm/pull/7386
